sync/src/synchronizer/block_fetcher.rs

```diff
@@ -172,6 +173,25 @@ impl BlockFetcher {
 
         // The headers in `fetch` may be unordered. Sort them by number.
         fetch.sort_by_key(|header| header.number());
-        Some(fetch.iter().map(core::HeaderView::hash).collect())
+
+        let tip = self.active_chain.tip_number();
```

Comment:
After the download window has been added back, an alternative way to check whether the download is stuck is checking fetch is empty and this peer inflight download queue is also empty. This is more efficient since if the peer has some works to do, we don't bother to fix the issue.

Reply:
It is not possible to judge the need for a mark based on fetch being empty. Because when in close proximity to the tip, fetch list will also always be empty